core/nightshadev2/src/nightshade.rs

```diff
@@ -0,0 +1,602 @@
+/// Nightshade v2
+use std::cmp::{max, min};
+use std::cmp::Ordering;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashSet;
+use std::hash::{Hash, Hasher};
+
+pub type AuthorityId = usize;
+pub type BLSSignature = u64;
+pub type NSSignature = usize;
+pub type NSHash = u64;
+
+const COMMIT_THRESHOLD: i64 = 3;
+
+pub enum NSResult {
+    Updated(Option<State>),
+    Error(String),
+}
+
+/// Nightshade consensus run on top of outcomes proposed by each authority.
+/// Blocks represent authorities proposal.
+#[derive(Debug, Clone, Serialize)]
+pub struct Block<P> {
+    pub header: BlockHeader,
+    payload: P,
+}
+
+impl<P: Hash> Block<P> {
+    pub fn new(author: AuthorityId, payload: P) -> Self {
+        // TODO: Use custom hasher
+        let mut hasher = DefaultHasher::new();
+        payload.hash(&mut hasher);
+        let hash = hasher.finish();
+
+        Self {
+            header: BlockHeader {
+                author,
+                hash,
+            },
+            payload,
+        }
+    }
+
+    /// Authority proposing the block
+    pub fn author(&self) -> AuthorityId {
+        self.header.author
+    }
+
+    /// Hash of the payload contained in the block
+    pub fn hash(&self) -> NSHash {
+        self.header.hash
+    }
+}
+
+/// BlockHeaders are used instead of Blocks as authorities proposal in the consensus.
+/// They are used to avoid receiving two different proposals from the same authority,
+/// and penalize such behavior.
+#[derive(Debug, Clone, Serialize, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockHeader {
+    /// Authority proposing the block.
+    pub author: AuthorityId,
+    /// Hash of the payload contained in the block.
+    hash: NSHash,
+}
+
+
+/// Triplet that describe the state of each authority in the consensus.
+///
+/// Notes:
+/// We are running consensus on authorities rather than on outcomes, `endorses` refers to an authority.
+/// "outcome" will be used instead of "authority" to avoid confusion.
+///
+/// The order of the fields are very important since lexicographical comparison is used derived from `PartialEq`.
+#[derive(Debug, Clone, Serialize, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BareState {
+    /// How much confidence we have on `endorses`.
+    confidence0: i64,
```

Comment:
primary_confidence / secondary_confidence?
or something like this?

Reply:
primary/secondary sounds good. Done